File: 백준/Silver/2961. 도영이가 만든 맛있는 음식/도영이가 만든 맛있는 음식.py
import sys
input = sys.stdin.readline
import itertools

n = int(input())
tastes = [tuple(map(int, input().split())) for _ in range(n)]
result = 2e9

for i in range(1, n+1):
    for combis in itertools.combinations(tastes, i):
        sour = 1
        bitter = 0
        for combi in combis:
            s, b = combi
            sour *= s
            bitter += b
        if abs(sour - bitter) < result:
            result = abs(sour - bitter)

print(result)

File: 백준/Silver/2961. 도영이가 만든 맛있는 음식/도영이가 만든 맛있는 음식.java
import java.io.BufferedReader;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.HashMap;
import java.util.StringTokenizer;

public class Main {
	static int N;
	static int[][] tastes;
	
	static int findRecipe() {
		int res = Integer.MAX_VALUE;
		
		for (int i = 1; i < (1 << N); i++) {
			int sour = 1;
			int bitter = 0;
			
			for (int j = 0; j < N; j++) {
				if ((i & (1 << j)) != 0) {
					sour *= tastes[j][0];
					bitter += tastes[j][1];
				}
			}
			res = Math.min(res, Math.abs(sour - bitter));
		}
		return res;
	}
	
	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		N = Integer.parseInt(br.readLine());
		tastes = new int[N][2];
		
		for (int i = 0; i < N; i++) {
			StringTokenizer numbers = new StringTokenizer(br.readLine());
			int[] taste = {Integer.parseInt(numbers.nextToken()), Integer.parseInt(numbers.nextToken())};
			tastes[i] = taste;
		}
		System.out.println(findRecipe());
	}

}